[SPARK-44198][CORE] Support propagation of the log level to the executors

vinodkc opened the pull request:
### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Currently, the **sc.setLogLevel()** method only sets the log level on the Spark driver, failing to reflect the desired log level on the executors. With  _--conf **spark.log.level**_  or **sc.setLogLevel()**, spark allows tuning the log level in the driver process, but it is not reflecting the log level on executors.

### Why are the changes needed?
This inconsistency can lead to difficulties in debugging and monitoring Spark applications, as log messages from the executors may not align with the expected log level set on the user code.
This PR aims to propagate the log level changes to executors when  sc.setLogLevel()  is called or send the current log level when a new executor is getting registered


<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
No, but with this PR, both driver and executor will show same log level
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Tested manually to verify the same log levels on both driver and executor

vinodkc committed:
Propagate log level to executors

vinodkc commented:
CC @dongjoon-hyun

dongjoon-hyun reviewed (commented):
I'm not sure this is consistent with the original PR's proposal, SPARK-43782 . Let's get comments.

Gentle ping, @jwang0306 , @grundprinzip , @LuciferYang , @HyukjinKwon

LuciferYang commented:
> I'm not sure this is consistent with the original PR's proposal
+1

Firstly, I don't think the log level of the driver and executor must be consistent. 

But If this ability is needed, I suggest adding a flag to decide whether to propagate the log level to the executor

vinodkc commented:
@LuciferYang and @dongjoon-hyun , Thanks for the review comments.
Sure, I'll add a flag to decide whether to propagate the log level to the executor.

mridulm reviewed (commented):
Took a quick pass

vinodkc commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -383,20 +383,22 @@ class SparkContext(config: SparkConf) extends Logging {
     val upperCased = logLevel.toUpperCase(Locale.ROOT)
-    require(SparkContext.VALID_LOG_LEVELS.contains(upperCased),
+    require(
+      SparkContext.VALID_LOG_LEVELS.contains(upperCased),
```
./dev/scalafmt did the formatting, if required I can revert it.

mridulm replied:
Please revert

vinodkc replied:
Done

vinodkc commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -383,20 +383,22 @@ class SparkContext(config: SparkConf) extends Logging {
-        System.err.printf("Setting Spark log level to \"%s\".\n", level)
-      }
-      setLogLevel(level)
-    }
```
Moved this down to avoid an NPE with this PR, however, this initialization message will still be printed.

mridulm replied:
To clarify, I did not mean this specific message - but all the messages which would not get printed out due to log level not changing (until later).

vinodkc replied:
Done, moved the back to initial position

vinodkc commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -311,13 +311,24 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+          // propagate current log level to new executor only if flag is true
+          if (conf.get(EXECUTOR_ALLOW_SYNC_LOG_LEVEL)) {
+            data.executorEndpoint.send(RefreshExecutor(Map("logLevel" -> Utils.getLogLevel)))
+          }
```
The actual issue is, with `spark.log.level` in `sc.conf` or `--conf spark.log.level` or `sc.setLogLevel` ,  executor log level can not be changed. it just changes the Driver side  log level alone
To debug issues on the Executor side, only option currently available is to change the executor log level by modifying log4j/log4j2 properties file. This will force the user to make custom log4j/log4j2 files for each application. 
With this PR, the current log level applied through  `spark.log.leve`l in `sc.conf` or `--conf spark.log.level`  or `sc.setLogLevel` will be passed to executors

mridulm replied:
To clarify, I meant this change at this specific location alone - 
* Set log level in SparkContext.conf when initializing and/or when user calls `setLogLevel`
* When executors are initialized, they already pull the conf from driver - and at this point, apply the log level.

If/when user changes log level dynamically, that would still need to be propagated to executors (as executor(s) have already been initialized) - but we dont need to push it each time a new executor comes up: only when level is changed.

grundprinzip replied:
Conceptually, I agree with @mridulm. The only thing I'm wondering is if it creates some undesired side-effects if the user turns on the "sync" flag and then sets the log level to overwrite the executor configuration.

I don't think this is a real case though.

vinodkc replied:
@mridulm , Done, when executors are initialized, applied the log level of the driver in the executor.
@grundprinzip , If the user turns on the "sync", it will override the executor log configuration only if `spark.log.level`  is applied on `SparkContext` or `sc.setLogLevel()` explicitly called with a different log level than default log level configured on log4j2 , hence there won't be a side effect in that scenario.

vinodkc commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala:
```diff
@@ -49,6 +49,10 @@ private[spark] object CoarseGrainedClusterMessages {
 
+  case class RefreshExecutors(msg: Map[String, String]) extends CoarseGrainedClusterMessage
+
+  case class RefreshExecutor(msg: Map[String, String]) extends CoarseGrainedClusterMessage
```
I gave a generating name to this message as this can hold a map of message values and in the future,  this can be reused,  if any other message needs to be passed to the executor. 

 If that is not required, I can change this to a specific message name. 
eg: `UpdateExecutorsLogLevel(msg: String)`  & `UpdateExecutorLogLevel(msg: String)`
    OR 
`RefreshExecutorsLogLevel(msg: String)` & `RefreshExecutorLogLevel(msg: String)`

Please give your suggestions

mridulm replied:
Update sounds better to me.

grundprinzip replied:
+1 that the messasge name is too generic. It should reflect the intent of the caller. I would prefer `Update*` as well

vinodkc replied:
Changed to `UpdateExecutorsLogLevel(msg: String)` & `UpdateExecutorLogLevel(msg: String)`

vinodkc commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -638,6 +649,17 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+    try {
+      if (driverEndpoint != null) {
+        logInfo("Refreshing executors")
+        driverEndpoint.askSync[Boolean](RefreshExecutors(msg))
```
Done

vinodkc commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2464,16 +2464,27 @@ private[spark] object Utils extends Logging with SparkClassUtils {
     Logging.sparkShellThresholdLevel = null
   }
 
+  private def getLogContext = {
```
Done

mridulm commented:
+CC @dongjoon-hyun, @HyukjinKwon as well - you guys might be more familiar with the earlier change for connect

grundprinzip commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -384,20 +384,22 @@ class SparkContext(config: SparkConf) extends Logging {
+    // Update only if new log level is not same as current log level
+    if (upperCased != Utils.getLogLevel) {
+      Utils.setLogLevel(Level.toLevel(upperCased))
+    // Inform all executors about the change
```
nit: indent

vinodkc replied:
Done

grundprinzip commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2464,16 +2465,27 @@ private[spark] object Utils extends Logging with SparkClassUtils {
   }
 
+  private lazy val getLogContext: (LoggerContext, LoggerConfig) = {
+    val ctx = LogManager.getContext(false).asInstanceOf[LoggerContext]
```
Is this null safe?

vinodkc replied:
Yes, it is the refactor of existing code

grundprinzip reviewed (commented):
The title and the body of the commit message should reflect the fact that this does not fix the inconsistency per se, but is actually an enhancement that allows propagation of the log level to the executors if the new spark conf is provided.

There is a proposal by @mridulm to make the propagation to new executors based on the Spark conf rather than a message for new executors, but I'm wondering if this could lead to race conditions. I don't think its important enough to make this fully safe, but it would be great to add proper comment sin the appropriate places to help readers understand the intents and purposes.

LuciferYang commented:
Hmm.. I have a question(not just about this one): I am not sure if this will conflict with `automatic-reconfiguration` of log4j2. If the user has configured `spark.log.level|spark.executor.sync.log.level` in Spark side, and also specified `monitorInterval` in the log4j2 configuration file, will the log be automatically reconfiguration by log4j2 and making the spark configurations invalid?

vinodkc commented:
@LuciferYang , Thanks for the pointer. 
I just tested that scenario by adding `monitorInterval=5000` in `./conf/log4j2.properties`. 
Observation: Log4j2 will not do automatic reconfiguration as the log4j2.properties file is not modified from outside.

vinodkc commented:
> There is a proposal by @mridulm to make the propagation to new executors based on the Spark conf rather than a message for new executors, but I'm wondering if this could lead to race conditions.
During executor startup flow, this PR applies the log level, there won't be a race condition  from this change

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -318,6 +318,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         stop()
 
+      case UpdateExecutorsLogLevel(msg) =>
+        logInfo("Asking each executor to refresh")
```
```suggestion
        logInfo("Asking each executor to refresh the log level to $logLevel")
```

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -653,6 +660,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     }
   }
 
+  override def update(msg: String): Unit = {
```
```suggestion
  override def updateLogLevel(logLevel: String): Unit = {
```

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -653,6 +660,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
+  override def update(msg: String): Unit = {
+      if (driverEndpoint != null) {
+        logInfo("Refreshing executors")
```
```suggestion
```

attilapiros commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -387,7 +387,15 @@ class SparkContext(config: SparkConf) extends Logging {
         s" ${SparkContext.VALID_LOG_LEVELS.mkString(",")}")
-    Utils.setLogLevel(Level.toLevel(upperCased))
+    // Update only if new log level is not same as current log level
+    if (upperCased != Utils.getLogLevel) {
```
What if "spark.log.level" was not use but the executor log level is different from the driver based on the log4 configs? One might try to call `sc.setLogLevel` but because of this check the new log level might not reach the executor if he calls it with the current driver log level.

So I would move the following part outside of this outer if:
```
      if (conf.get(EXECUTOR_ALLOW_SYNC_LOG_LEVEL) && _schedulerBackend != null) {
        _conf.set(SPARK_LOG_LEVEL.key, upperCased)
        _schedulerBackend.update(upperCased)
      }
```

attilapiros commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -473,6 +476,12 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
+      if (driverConf.get(EXECUTOR_ALLOW_SYNC_LOG_LEVEL)) {
+        // Set the current log level of driver in Executor
+        driverConf.get(SPARK_LOG_LEVEL).foreach(l => updateLogLevel(l))
+
```
```suggestion
```

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -318,6 +318,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         context.reply(true)
         stop()
 
+      case UpdateExecutorsLogLevel(msg) =>
```
```suggestion
      case UpdateExecutorsLogLevel(logLevel) =>
```

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala:
```diff
@@ -49,6 +49,10 @@ private[spark] object CoarseGrainedClusterMessages {
   case class KillExecutorsOnHost(host: String)
     extends CoarseGrainedClusterMessage
 
+  case class UpdateExecutorsLogLevel(msg: String) extends CoarseGrainedClusterMessage
```
```suggestion
  case class UpdateExecutorsLogLevel(logLevel: String) extends CoarseGrainedClusterMessage
```

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala:
```diff
@@ -49,6 +49,10 @@ private[spark] object CoarseGrainedClusterMessages {
 
+  case class UpdateExecutorsLogLevel(msg: String) extends CoarseGrainedClusterMessage
+
+  case class UpdateExecutorLogLevel(msg: String) extends CoarseGrainedClusterMessage
```
```suggestion
  case class UpdateExecutorLogLevel(logLevel: String) extends CoarseGrainedClusterMessage
```

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -318,6 +318,13 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+      case UpdateExecutorsLogLevel(msg) =>
+        logInfo("Asking each executor to refresh")
+        for ((_, executorData) <- executorDataMap) {
+          executorData.executorEndpoint.send(UpdateExecutorLogLevel(msg))
```
```suggestion
          executorData.executorEndpoint.send(UpdateExecutorLogLevel(logLevel))
```

attilapiros reviewed (commented):
In the description you mention "Added a test case." but currently there is no new test added (neither was any modification so this is not tested even with the existing tests).

What happens when a new executor is created during the application run? 
Like in case of static allocation when one executor is lost and a new one is created instead of the old one? Or in case of dynamic allocation when Spark scales up in the number of executors? Does the executor started with the new log level? 

Can you show please attach a manual test output where the feature is tested at least with 3 log levels and with dynamic allocation, like
1) initial one: WARN (this can be from the log4j config)
2) first update to INFO (`sc.setLogLevel("INFO")`)
3) second update to DEBUG (`sc.setLogLevel("DEBUG")`, I am interested whether the config update really works in the 2nd time too)
4) scaling up by executing a higher number (backlogged) tasks as previously

vinodkc commented:
### Manual test steps where the feature is tested at least with 3 log levels. With new executor creation after sc.setLogLevel() call.
1) initial one: WARN (this can be from the log4j config)
`rootLogger.level = warn`
```
  > ./bin/spark-shell  --master yarn  --conf spark.executor.sync.log.level=true
```
2) first update to INFO (sc.setLogLevel("INFO"))
**a) In the driver** 

```
  scala> sc.setLogLevel("INFO")
  23/07/10 20:59:42 INFO [dispatcher-CoarseGrainedScheduler] YarnSchedulerBackend$YarnDriverEndpoint: Asking each executor to refresh the log level to INFO
  scala> sc.parallelize(1 to 10).count
23/07/10 21:00:07 INFO [main] SparkContext: Starting job: count at <console>:24
23/07/10 21:00:07 INFO [dag-scheduler-event-loop] DAGScheduler: Got job 0 (count at <console>:24) with 2 output partitions
23/07/10 21:00:07 INFO [dag-scheduler-event-loop] DAGScheduler: Final stage: ResultStage 0 (count at <console>:24)
```  


 **b) In the executor**

```
Container: container_1689037381071_0016_01_000002 on 10.0.0.223_50550
LogAggregationType: AGGREGATED
=====================================================================
LogType:stderr
LogLastModifiedTime:Mon Jul 10 21:01:40 -0700 2023
LogLength:6827
LogContents:
23/07/10 20:59:37 WARN [main] NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/07/10 21:00:07 INFO [dispatcher-Executor] YarnCoarseGrainedExecutorBackend: Got assigned task 0
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] Executor: Running task 0.0 in stage 0.0 (TID 0)
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] TorrentBroadcast: Started reading broadcast variable 0 with 1 pieces (est
imated total size 4.0 MiB)
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] TransportClientFactory: Successfully created connection to /10.0.0.223:55
625 after 1 ms (0 ms spent in bootstraps)
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] MemoryStore: Block broadcast_0_piece0 stored as bytes in memory (estimate
d size 1782.0 B, free 93.3 MiB)
```
3) second update to DEBUG (sc.setLogLevel("DEBUG"),to check whether the config update really works the 2nd time too)
```
sc.setLogLevel("DEBUG")
sc.parallelize(1 to 10).count
```
Driver printed DEBUG logs
 In the executor [container_1689037381071_0016_01_000002] log I could see DEBUG log.
_**This confirms that log Level was propagated to Executor**_ 
```
Container: container_1689037381071_0016_01_000002 on 10.0.0.223_50550
LogAggregationType: AGGREGATED
=====================================================================
LogType:stderr
LogLastModifiedTime:Mon Jul 10 21:01:40 -0700 2023
LogLength:6827
LogContents:
23/07/10 20:59:37 WARN [main] NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/07/10 21:00:07 INFO [dispatcher-Executor] YarnCoarseGrainedExecutorBackend: Got assigned task 0
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] Executor: Running task 0.0 in stage 0.0 (TID 0)
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] TorrentBroadcast: Started reading broadcast variable 0 with 1 pieces (estimated total size 4.0 MiB)
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] TransportClientFactory: Successfully created connection to /****:55
625 after 1 ms (0 ms spent in bootstraps)
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] MemoryStore: Block broadcast_0_piece0 stored as bytes in memory (estimate
d size 1782.0 B, free 93.3 MiB)
23/07/10 21:00:07 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] TorrentBroadcast: Reading broadcast variable 0 took 128 ms
23/07/10 21:00:08 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] MemoryStore: Block broadcast_0 stored as values in memory (estimated size
 2.8 KiB, free 93.3 MiB)
23/07/10 21:00:08 INFO [Executor task launch worker for task 0.0 in stage 0.0 (TID 0)] Executor: Finished task 0.0 in stage 0.0 (TID 0). 973 bytes result sent t
o driver
23/07/10 21:00:08 INFO [dispatcher-Executor] YarnCoarseGrainedExecutorBackend: Got assigned task 1
23/07/10 21:00:08 INFO [Executor task launch worker for task 1.0 in stage 0.0 (TID 1)] Executor: Running task 1.0 in stage 0.0 (TID 1)
23/07/10 21:00:08 INFO [Executor task launch worker for task 1.0 in stage 0.0 (TID 1)] Executor: Finished task 1.0 in stage 0.0 (TID 1). 887 bytes result sent t
o driver
23/07/10 21:00:40 INFO [dispatcher-Executor] YarnCoarseGrainedExecutorBackend: Got assigned task 2
23/07/10 21:00:40 INFO [Executor task launch worker for task 0.0 in stage 1.0 (TID 2)] Executor: Running task 0.0 in stage 1.0 (TID 2)
23/07/10 21:00:40 DEBUG [Executor task launch worker for task 0.0 in stage 1.0 (TID 2)] Executor: task 0.0 in stage 1.0 (TID 2)'s epoch is 0
```
4) Scale down the executor by killing container container_1689037381071_0016_01_000002. (Instead of scaling up, did scale down to zero executors).In that scenario also new executors will be created to run the next jobs
Killed current executor container_1689037381071_0016_01_000002
```
>yarn container -signal container_1689037381071_0016_01_000002  GRACEFUL_SHUTDOWN
```
On spark-shell run a new job
```
sc.parallelize(1 to 10).count
```
AM created a new  Container: `container_1689037381071_0016_01_000003`
 In the **new executor** [container_1689037381071_0016_01_000003] log, we can see DEBUG log. _**This confirms that newly create Executor also  pulled the latest loglevel from Driver**_
```
Container: container_1689037381071_0016_01_000003 on 10.0.0.223_50550
LogAggregationType: AGGREGATED
=====================================================================
LogType:stderr
LogLastModifiedTime:Mon Jul 10 21:01:40 -0700 2023
LogLength:10269
LogContents:
23/07/10 21:01:12 WARN [main] NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/07/10 21:01:13 INFO [main] SecurityManager: Changing view acls to: vkc
23/07/10 21:01:13 INFO [main] SecurityManager: Changing modify acls to: vkc
23/07/10 21:01:13 INFO [main] SecurityManager: Changing view acls groups to:
23/07/10 21:01:13 INFO [main] SecurityManager: Changing modify acls groups to:
.......
23/07/10 21:01:13 DEBUG [main] DiskBlockManager: Adding shutdown hook
23/07/10 21:01:13 DEBUG [main] ShutdownHookManager: Adding shutdown hook
23/07/10 21:01:13 INFO [main] MemoryStore: MemoryStore started with capacity 93.3 MiB
23/07/10 21:01:13 INFO [dispatcher-Executor] YarnCoarseGrainedExecutorBackend: Connecting to driver: spark://CoarseGrainedScheduler@******:55599
23/07/10 21:01:13 DEBUG [dispatcher-Executor] YarnCoarseGrainedExecutorBackend: Resource profile id is: 0
23/07/10 21:01:13 INFO [dispatcher-Executor] ResourceUtils: ==============================================================
23/07/10 21:01:13 INFO [dispatcher-Executor] ResourceUtils: No custom resources configured for spark.executor.
23/07/10 21:01:13 INFO [dispatcher-Executor] ResourceUtils: ==============================================================
23/07/10 21:01:13 DEBUG [rpc-client-3-1] YarnContainerInfoHelper: Base URL for logs: http://10.0.0.223:8042/node/containerlogs/container_1689037381071_0016_01_000003/***
23/07/10 21:01:13 INFO [dispatcher-Executor] YarnCoarseGrainedExecutorBackend: Successfully registered with driver
23/07/10 21:01:13 INFO [dispatcher-Executor] Executor: Starting executor ID 2 on host *****
```

vinodkc commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -585,6 +593,12 @@ class SparkContext(config: SparkConf) extends Logging {
+    // After initialisation of  _schedulerBackend, if EXECUTOR_ALLOW_SYNC_LOG_LEVEL is true and
+    // SPARK_LOG_LEVEL is already set, sync the new log level with all the current executors.
+    if (_conf.get(EXECUTOR_ALLOW_SYNC_LOG_LEVEL)) {
+      _conf.get(SPARK_LOG_LEVEL).foreach( logLevel => _schedulerBackend.updateLogLevel(logLevel))
```
Done

vinodkc commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -653,6 +660,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+  override def updateLogLevel(logLevel: String): Unit = {
+      if (driverEndpoint != null) {
+        driverEndpoint.ask[Boolean](UpdateExecutorsLogLevel(logLevel))
+      }
```
Done

vinodkc commented:
> In the description you mention "Added a test case." but currently there is no new test added (neither was any modification so this is not tested even with the existing tests).
In fact, I had added a test case, however, based on a review comment, I had to change the implementation of the driver to executor log propagation and hence remove that test case as it was irrelevant after the code change. Though I forgot to update the Test section of the PR comment. 
Updated that Description now.

vinodkc commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -143,9 +143,10 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+    // Spark configuration sent to executors. This is a method so that subclasses of the
+    // scheduler can modify the SparkConf object and latest spark configurations
+    // shall be sent to executors.
+    private def sparkProperties = scheduler.sc.conf.getAll
```
Change `lazy val` to `def` to get the latest spark configurations. Will `lazy val`, new executor gets a snapshot of original spark conf and that will cause propagation of the initial log level to new executors.
To avoid that scenario, sparkProperties should not be a lazy val.

attilapiros replied:
This change has some performance disadvantages. What about storing the last `logLevel` received via `UpdateExecutorsLogLevel`?

vinodkc replied:
sure, I'll fix it without altering this `lazy val`

attilapiros replied:
You needed this `lazy val` => `def` change because you found a bug during the execution of the manual testing, right? In this case when you ready with the change please redo the testing.

vinodkc replied:
Update the code and manually  tested

attilapiros commented:
I do not like the idea of using a spark config for a dynamically changing value like the updated log level. 

So I would store the the updated log level in `CoarseGrainedSchedulerBackend` in a member var (`Option[String]`) update it whenever a `UpdateExecutorsLogLevel` arrives and pass that to the executor backends in a new member val of the `SparkAppConfig`.

vinodkc committed:
store log level in CoarseGrainedSchedulerBackend

vinodkc commented:
@attilapiros , Thanks for the review comments, I updated the code to store log level in `CoarseGrainedSchedulerBackend` and then  pass that to the executor backends in a new member val of the `SparkAppConfig`

attilapiros commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -473,6 +476,11 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
 
       driverConf.set(EXECUTOR_ID, arguments.executorId)
+
+      if (driverConf.get(EXECUTOR_ALLOW_SYNC_LOG_LEVEL)) {
+        // Set the current log level of driver in Executor
+        cfg.logLevel.foreach(l => updateLogLevel(l))
+      }
```
```suggestion
      cfg.logLevel.foreach(l => updateLogLevel(l))
```
I think the check is not needed as when `EXECUTOR_ALLOW_SYNC_LOG_LEVEL` is false then the `cfg.logLevel` will be `None`.

attilapiros commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -558,6 +566,11 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
       resourcesFileOpt, resourceProfileId)
   }
 
+  private def updateLogLevel(newLogLevel: String): Unit = {
+    if (newLogLevel != Utils.getLogLevel) {
+      Utils.setLogLevel(Level.toLevel(newLogLevel))
+    }
+  }
```
rename to `setLogLevelIfNeeded` and move it to `Utils` you can then use it in `SparkContext` too

vinodkc committed:
Fix code review comments

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -653,6 +667,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     }
   }
 
+  override def updateLogLevel(logLevel: String): Unit = {
```
Nit: 
```suggestion
  override def updateExecutorsLogLevel(logLevel: String): Unit = {
```
Also update this please in the base class and  in the SparkContext where it is used.

attilapiros commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2436,16 +2437,35 @@ private[spark] object Utils extends Logging with SparkClassUtils {
 
+
+  def setLogLevelIfNeeded(newLogLevel: String): Unit = {
+    // Update only if new log level is not same as current log level
```
Nit: remove the comment (we only need comments for the complex code parts)

attilapiros commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -473,6 +475,10 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
       driverConf.set(EXECUTOR_ID, arguments.executorId)
+
+      // Set the current log level of driver in Executor
+      cfg.logLevel.foreach(l => Utils.setLogLevelIfNeeded(l))
```
```suggestion
      cfg.logLevel.foreach(logLevel => Utils.setLogLevelIfNeeded(logLevel))
```

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -123,6 +123,9 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   @volatile protected var currentExecutorIdCounter = 0
 
+  // Current log level of driver to send to executor
+  @volatile private var logLevel: Option[String] = None
```
Nit: rename to `currentLogLevel`

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -345,7 +358,8 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
           Option(delegationTokens.get()),
-          rp)
+          rp,
+          CoarseGrainedSchedulerBackend.this.logLevel)
```
Nit: I do not think we need "CoarseGrainedSchedulerBackend.this." here, do we?

vinodkc replied:
Not needed. Fixed

attilapiros reviewed (commented):
We are very close.

vinodkc committed:
Fixed review comments

attilapiros reviewed (approved):
LGTM 
I will wait one week (till 27th of July) and if it gets no new comments / change request I will merge this

LuciferYang commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -2244,6 +2244,14 @@ package object config {
+    ConfigBuilder("spark.executor.sync.log.level")
+      .doc("If set to true, log level applied through SparkContext.setLogLevel() method " +
+        "will be propagated to all executors.")
+      .version("3.5.0")
```
@attilapiros Will this be merged into Spark 3.5.0? The current development branch is already at 4.0.0.

vinodkc replied:
@LuciferYang , Updated to `4.0.0`.

vinodkc committed:
updated version

pan3793 commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -2244,6 +2244,14 @@ package object config {
       .createWithDefault(false)
 
+  private[spark] val EXECUTOR_ALLOW_SYNC_LOG_LEVEL =
+    ConfigBuilder("spark.executor.sync.log.level")
```
We should be careful to use `.` to introduce the unnecessary configuration namespace, suggested alternatives are

- `spark.executor.syncLogLevel`
- `spark.executor.syncLogLevel.enabled`

vinodkc replied:
Thanks.
Changed to `spark.executor.syncLogLevel.enabled`

vinodkc committed:
modified property name

vinodkc committed:
Merge with master

srowen commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala:
```diff
@@ -49,6 +50,10 @@ private[spark] object CoarseGrainedClusterMessages {
   case class KillExecutorsOnHost(host: String)
     extends CoarseGrainedClusterMessage
 
+  case class UpdateExecutorsLogLevel(logLevel: String) extends CoarseGrainedClusterMessage
```
Looks like this might have caused mima issues in Scala 2.13: https://github.com/apache/spark/pull/42358#pullrequestreview-1563933815

LuciferYang replied:
If we need the mima check pass for both Scala 2.12 and Scala 2.13, `SparkAppConfig` might need to be designed as follows:

```scala
  case class SparkAppConfig(
      sparkProperties: Seq[(String, String)],
      ioEncryptionKey: Option[Array[Byte]],
      hadoopDelegationCreds: Option[Array[Byte]],
      resourceProfile: ResourceProfile)
    extends CoarseGrainedClusterMessage {

    private var _logLevel: Option[String] = None

    def logLevel: Option[String] = _logLevel

    def this(
        sparkProperties: Seq[(String, String)],
        ioEncryptionKey: Option[Array[Byte]],
        hadoopDelegationCreds: Option[Array[Byte]],
        resourceProfile: ResourceProfile,
        logLevel: Option[String]) = {
      this(sparkProperties, ioEncryptionKey, hadoopDelegationCreds, resourceProfile)
      this._logLevel = logLevel
    }
  }
```

srowen replied:
This is an internal class right? if so, ignoring it is OK

LuciferYang replied:
OK, let's just ignore it

attilapiros replied:
Thanks for fixing this!
